### ISSUE TITLE
 Getting NoClassDefFoundError: javax/security/jacc/EJBMethodPermission with native builds 

### ISSUE THREAD

**Issue by theX55** [quarkiverse/quarkus-google-cloud-services#147]
Hi all,

we're facing currently the issue, that our native build crashes, with the error NoClassDefFoundError: javax/security/jacc/EJBMethodPermission.

What we use:
Maven 3.8.1
Quarkus 1.13.7 (also tried with 1.13.0)
GraalVM 21.1.0 (also tried 21.0.0.2 & 20.3.1)
quarkus-google-cloud-pubsub 0.7.0 (also tried 0.8.0 & 0.6.0) **<--- if we leave this dependecy the build is green**

**--> Rel:**  Getting NoClassDefFoundError with native builds #110 

Here is the stacktrace of our build:

> Status: Downloaded newer image for quay.io/quarkus/ubi-quarkus-native-image:21.0.0-java11
> quay.io/quarkus/ubi-quarkus-native-image:21.0.0-java11
> [INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] Running Quarkus native-image plugin on GraalVM Version 21.0.0 (Java Version 11.0.10+8-jvmci-21.0-b06)
> [INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildRunner] docker run --env LANG=C --user 0:0 --rm -v /opt/atlassian/pipelines/agent/build/target/authorization-service-1.0.0-SNAPSHOT-native-image-source-jar:/project:z quay.io/quarkus/ubi-quarkus-native-image:21.0.0-java11 -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-DCoordinatorEnvironmentBean.transactionStatusManagerEnable=false -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 -J-Duser.language=en -J-Duser.country= -J-Dfile.encoding=UTF-8 --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy\$BySpaceAndTime -H:+JNI -H:+AllowFoldMethods -jar authorization-service-1.0.0-SNAPSHOT-runner.jar -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -J-Xmx6g -H:-AddAllCharsets -H:EnableURLProtocols=http,https --enable-all-security-services --no-server -H:-UseServiceLoaderFeature -H:+StackTrace authorization-service-1.0.0-SNAPSHOT-runner
> [authorization-service-1.0.0-SNAPSHOT-runner:25]    classlist:   8,987.37 ms,  1.20 GB
> [authorization-service-1.0.0-SNAPSHOT-runner:25]        setup:     782.91 ms,  1.20 GB
> Fatal error:java.lang.NoClassDefFoundError
> 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
> 	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
> 	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
> 	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:603)
> 	at java.base/java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1006)
> 	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:488)
> 	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:370)
> 	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:529)
> 	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:119)
> 	at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:561)
> Caused by: java.lang.NoClassDefFoundError: javax/security/jacc/PolicyContextException
> 	at java.base/java.lang.Class.getDeclaredMethods0(Native Method)
> 	at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3166)
> 	at java.base/java.lang.Class.getDeclaredMethods(Class.java:2309)
> 	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.handleDeletedClass(AnnotationSubstitutionProcessor.java:512)
> 	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.handleClass(AnnotationSubstitutionProcessor.java:305)
> 	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.init(AnnotationSubstitutionProcessor.java:265)
> 	at com.oracle.svm.hosted.NativeImageGenerator.createDeclarativeSubstitutionProcessor(NativeImageGenerator.java:929)
> 	at com.oracle.svm.hosted.NativeImageGenerator.setupNativeImage(NativeImageGenerator.java:863)
> 	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:561)
> 	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:476)
> 	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
> 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
> 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
> 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
> 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
> 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
> Caused by: java.lang.ClassNotFoundException: javax.security.jacc.PolicyContextException
> 	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:471)
> 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
> 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
> 	... 16 more
> Error: Image build request failed with exit status 1
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  01:02 min
> [INFO] Finished at: 2021-06-24T15:53:03Z
> [INFO] ------------------------------------------------------------------------
> [ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:1.13.0.Final:build (default) on project authorization-service: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
> [ERROR] 	[error]: Build step io.quarkus.deployment.pkg.steps.NativeImageBuildStep#build threw an exception: java.lang.RuntimeException: Failed to build native image
> [ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.build(NativeImageBuildStep.java:209)
> [ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> [ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> [ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> [ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> [ERROR] 	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:920)
> [ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
> [ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
> [ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
> [ERROR] 	at java.base/java.lang.Thread.run(Thread.java:834)
> [ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
> [ERROR] Caused by: java.lang.RuntimeException: Image generation failed. Exit code: 1
> [ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.imageGenerationFailed(NativeImageBuildStep.java:350)
> [ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.build(NativeImageBuildStep.java:181)
> [ERROR] 	... 10 more
> [ERROR] -> [Help 1]

Would be great if you've an idea what's casing that problem.

Thank you!

Best regards
Manuel

**Comment by loicmathieu** [comment]
Hi, thanks for reporting, we indead already have the issue reported but the reporter never provide a reproducer and our native buil in CI is working.

Can you provide a reproducer, as a GithHub repo or as a zip file attached in the issue ?

**Comment by theX55** [comment]
Hi @loicmathieu ,
we did some research on this issue. It seems, that the pubsub and the JDBC dependencies are not working proper together.
I created a example project also some pipeline runs.

Please have a look at: bitbucket repo`https://bitbucket.org/devdechea/example-pub-sub/src/master/`

**Comment by loicmathieu** [comment]
Thanks, I'll have a look at this one in a few weeks (vacation) if nobody else pick it before.

**Comment by loicmathieu** [comment]
@theX55 I dig a little on this issue and Hibernate ORM and pubsub extensions are not working together. I'll try to understand what's going on, my current asumption is that  some dependency from the Hibernate ORM trigger some incompatibility code path on pubsub.

**Comment by theX55** [comment]
Hi @loicmathieu,
I can assure, that Hibernate and PubSub can work together. We've made it work in the meantime. 
Sorry I forgot to replay here.
We solved the issue with adding this dependencies:
`
        <dependency>
            <groupId>org.jboss.spec.javax.security.jacc</groupId>
            <artifactId>jboss-jacc-api_1.5_spec</artifactId>
            <version>2.0.0.Final</version>
        </dependency>
        <dependency>
            <groupId>net.java.dev.jna</groupId>
            <artifactId>jna-platform</artifactId>
            <version>5.8.0</version>
        </dependency>`
I hope this can help you find the issue better or find a better solution.

**Comment by loicmathieu** [comment]
@dzou adding both PubSub and Hibernate triggers a GraalVM issue due to the missing ` javax/security/jacc/PolicyContextException`. The issue could be reproducible easily by creating a Quarkus apps with pubsub and Hibernate ORM (from code.quarkus.io for ex). Adding the dependencies from the previous comment fixes the issue.
Do you know if pubsub could trigger the inclusion of JACC in a certain way ? Quarkus Hibernate ORM extension disable JACC on itself.

**Comment by dzou** [comment]
Hmm, I think the issue is actually coming from our project `google-cloud-graalvm-support`. If I remove the `pubsub` dependency but add `google-cloud-graalvm-support`, it causes the issue. Then removing the dependency fixes it. However, adding the core `pubsub` library does not cause the issue.

Is there something that we need to be doing in [our project](https://github.com/GoogleCloudPlatform/google-cloud-graalvm-support/tree/master/google-cloud-graalvm-support) to work around this? It is a relatively small project so I could try to remove some things incrementally to find out what is the root cause. Curious if you know what we might be doing to trigger the inclusion though.

**Comment by loicmathieu** [comment]
@dzou I had a quick look at the library and didn't see what triggers the inclusion of `javax/security/jacc/PolicyContextException`.

The only thing I see is that we're still using an old version of the Google Cloud Library BOM, 16.1.0, but the native support library defines compatibility to version 20.8.0 or later. I can try upgrading to a more recent version and see if it makes any difference but this didn't seems related.

**Comment by dzou** [comment]
I think the issue might be stemming from the usage of `--report-unsupported-elements-at-runtime` (seems like it's another problem with the flag in reference to the issue we discussed earlier).

I have a PR ready to remove it but it got delayed due to the GraalVM `21.2.0` problems we had. Let me try this again.

**Comment by loicmathieu** [comment]
Thanks @dzou Quarkus Hibernate extension remove jacc usage so it may collide whith `--report-unsupported-elements-at-runtime` and the substitutions used. If you already have a PR this would be great as it causes issue on multiple areas.

**Comment by dzou** [comment]
Hey we just released [version 0.6.0](https://github.com/GoogleCloudPlatform/google-cloud-graalvm-support/releases/tag/0.6.0) of `google-cloud-graalvm-support` which I think should resolve this issue.

In your reproducer app, I overrided the version of `google-cloud-graalvm-support` by adding this dependency for version 0.6.0:

```
        <dependency>
            <groupId>com.google.cloud</groupId>
            <artifactId>google-cloud-graalvm-support</artifactId>
            <version>0.6.0</version>
        </dependency>
```

And this was able to get the build to work.

**Comment by loicmathieu** [comment]
Thanks @dzou, I'll check on the other issue that references this native image flag suppression.

I'm closing the one as #166 fixed it.
